reana_commons/gherkin_parser/parser.py

```diff
@@ -0,0 +1,259 @@
+        for step in scenario["steps"]:
+            steps.append((step["type"].strip(), _get_step_text(step)))
+    return steps
+
+
+def map_steps_to_functions(steps: list, data_fetcher: DataFetcherBase):
+    """Map each step to the corresponding step definition (function).
+
+    :param steps: A list of tuples, each containing the step type, and the step text.
+    :return: A dictionary mapping each step to the corresponding step definition (function).
+            Example:
+            {
+                "Action": {
+                    "The workflow status is finished": {
+                        "function": <function check_workflow_status at 0x7f7f7f7f7f7f>,
+                        "arguments": {"status_workflow": "finished"}
+                    }
+                },
+                ...,
+            }
+    :raise StepDefinitionNotFound: If the feature file contains a step for which no step definition is found.
+    """
+    step_mapping = {"Context": {}, "Action": {}, "Outcome": {}}
+    for step_type, step_text in steps:
+        found = False
+        # Get the list of all step definitions, divided by step type.
+        step_definitions = _get_step_definition_lists(data_fetcher)
+        for func in step_definitions[step_type]:
+            # Check if the step text matches any of the patterns.
+            parse_results = [
+                parse(pattern, step_text) for pattern in func._step_pattern
+            ]
+            result = next((r for r in parse_results if r is not None), None)
+            # Check if there was at least one non-None result.
+            if result is not None:
+                step_mapping[step_type][step_text] = {}
+                step_mapping[step_type][step_text]["function"] = func
+                step_mapping[step_type][step_text]["arguments"] = result.named
+                found = True
+                break
+        if not found:
+            logging.error(f"No step definition found for step: {step_text}")
+            raise StepDefinitionNotFound(
+                f"No step definition found for step: {step_text}"
+            )
+    return step_mapping
+
+
+def run_tests(
+    workflow: str,
+    feature_name: str,
+    feature_file,
+    step_mapping: Dict,
+) -> List[TestResult]:
+    """Run all the tests in the parsed feature file.
+
+    :param feature_name: The name of the feature inside the feature file.
+    :param workflow: The name of the workflow in REANA
+    :param feature_file: The parsed and compiled feature file, as returned by `parse_feature_file`.
+    :param step_mapping: A dictionary mapping each step to the corresponding step definition (function).
+    :return: A list of dictionaries, each corresponding to one scenario of the feature file.
+    """
+    test_results = []
+    for scenario in feature_file:
+        logging.info(f"Running scenario: {scenario['name']}...")
+        result = AnalysisTestStatus.passed
+        failed_testcase = ""
+        error_log = ""
```

Comment:
Any reason why `failed_testcase` and `error_log` were changed from `None` to `""`?

Reply:
Perhaps I should have mentioned it my commit, but this was done to make the server side version simpler (also from my testing, it doesn't cause any issues with continuous-reuse).

The rest api we're using doesn't have an easy way to specify both string and None return types simultaneously (passing tests were returning None for these two fields). I mentioned this to Tibor in a standup, and he recommended just returning the empty string.

Reply:
Although, having typed this out, I realise we could just check for None types before returning a response from the server...

Reply:
Yeah, I would not change things here, to avoid changing the behaviour of continuous-reuse. In any case, in the OpenAPI spec you can specify that a field can be nullable with `x-nullable`, see for example: https://github.com/reanahub/reana-server/blob/3596ee058a9888d723a8fcd6247bb14ff1095821/reana_server/rest/workflows.py#L3026

Reply:
I agree, changed back :)